### PR TITLE
Prog to move VMs

### PR DESCRIPTION
**Pull request by pykello**
This can be invoked using:

```ruby

vm.incr_stop
vm.incr_prepare_to_move
st = Prog::Vm::Metal::MoveVm.assemble(vm, vmh)
```

### CODE REVIEW

File: migrate/20260716_remote_storage_server.rb [hunk 1]
```diff
@@ -0,0 +1,17 @@
+# frozen_string_literal: true
+
+Sequel.migration do
+  change do
+    create_table(:remote_storage_server) do
+      column :id, :uuid, primary_key: true
```
Comment (jeremyevans) [inline migrate/20260716_remote_storage_server.rb]: Add a default using `gen_random_ubid_uuid` (search existing migrations for examples).

File: model/metal/vm.rb [hunk 2]
```diff
@@ -153,10 +153,24 @@ def storage_volumes
...
+    # Descriptor for a boot volume whose stripes come from a remote storage
+    # server. The PSK is wrapped with the volume's KEK, mirroring how archive
+    # credentials are delivered, so it is never sent to the host in plaintext.
+    def storage_remote_source(sv)
+      server = sv.remote_storage_server
+      {
+        "address" => server.address,
+        "psk_identity" => server.psk_identity,
+        "encrypted_psk" => sv.key_encryption_key_1.encrypt(Base64.decode64(server.psk), "remote-psk"),
+        "autofetch" => false,
```
Comment (jeremyevans) [inline model/metal/vm.rb]: The eventual goal is to move the VM storage from one host to another and then stop the remote storage server. I'm not sure how `autofetch` works, but ideally I think we would want:

1) As soon as possible, start fetching chunks sequentially from the remote storage server (running on original host) to current host.
2) If there is a request for a chunk that is not already loaded, prioritize that chunk request before other chunk requests.
3) After all chunks have been transferred, shut down the remote storage server.

File: prog/storage/remote_storage_server/nexus.rb [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+require "securerandom"
+
+# Runs an ubiblk remote-stripe-server daemon on the host of a source volume,
+# serving that volume over the remote stripe protocol (TLS-PSK) so another host
+# can boot a VM from it. The daemon always uses the v0.5.0 backend, which can
+# serve volumes created by older backends too.
+class Prog::Storage::RemoteStorageServer::Nexus < Prog::Base
+  subject_is :remote_storage_server
+
+  PORT_RANGE = (5500..5999)
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb:11]: Not a bit deal, but the port used in the specs isn't in this range :)

File: prog/storage/remote_storage_server/nexus.rb [hunk 4]
```diff
@@ -0,0 +1,118 @@
+# frozen_string_literal: true
+
+require "securerandom"
+
+# Runs an ubiblk remote-stripe-server daemon on the host of a source volume,
+# serving that volume over the remote stripe protocol (TLS-PSK) so another host
+# can boot a VM from it. The daemon always uses the v0.5.0 backend, which can
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb]: I think comments mentioning a specific version are likely to get out of date as new versions are released. I would remove the versions from the comments.

File: prog/storage/remote_storage_server/nexus.rb [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+  # Given the volume to serve, figure out its host, pick a free port on that
+  # host, mint a PSK, and start the server.
+  def self.assemble(vm_storage_volume_id)
+    source_volume = VmStorageVolume[vm_storage_volume_id]
+    fail "No existing VmStorageVolume" unless source_volume
+    fail "Source volume must be encrypted" unless source_volume.key_encryption_key_1
+
+    DB.transaction do
+      port = free_port(source_volume.vm.vm_host_id)
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb]: This assumes a port not used by an existing remote storage server will be available for use. How should we handle cases where it isn't? Maybe we can try that port, and if it isn't available, try a different port, and if that works, updating the port in the database?

File: prog/storage/remote_storage_server/nexus.rb [hunk 6]
```diff
@@ -0,0 +1,118 @@
...
+      Strand.create_with_id(id, prog: "Storage::RemoteStorageServer::Nexus", label: "start")
+    end
+  end
+
+  # The lowest port in PORT_RANGE not already used by a remote storage server on
+  # the same host.
+  def self.free_port(vm_host_id)
+    used = RemoteStorageServer
+      .join(:vm_storage_volume, id: :source_vm_storage_volume_id)
+      .join(:vm, id: Sequel[:vm_storage_volume][:vm_id])
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb]: Sequel will qualify the hash keys here to the last joined table by default, so I think this should work:

```suggestion
      .join(:vm, id: :vm_id)
```

Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb]: Alternatively, I think you can simplify this using `association_join`:

```suggestion
      .association_join(vm_storage_volume: :vm)
```

File: prog/storage/remote_storage_server/nexus.rb [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+      .where(Sequel[:vm][:vm_host_id] => vm_host_id)
+      .select_map(Sequel[:remote_storage_server][:port])
+    PORT_RANGE.find { |port| !used.include?(port) } || fail("No free port for remote storage server")
+  end
+
+  def before_run
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb]: This doesn't call `super`, which may be intentional, but I would add a comment if so.

File: prog/storage/remote_storage_server/nexus.rb [hunk 8]
```diff
@@ -0,0 +1,118 @@
...
+  end
+
+  label def start
+    register_deadline("wait", 5 * 60)
+    # The source volume's vhost backend keeps running when its VM is stopped, so
+    # it still holds the disk and the kek pipe. Stop it so the remote server can
+    # open the volume exclusively. The source VM cannot run again until its
+    # storage is restarted, which is the intended migration semantics.
+    source = remote_storage_server.source_vm_storage_volume
+    sshable.cmd("sudo systemctl stop :unit", unit: source.vhost_backend_systemd_unit_name)
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb:60]: In addition to stopping the storage, we may want to put the `Vm` in `admin_stop`, which prevents the `user` from starting the VM after it is stopped.

File: prog/storage/remote_storage_server/nexus.rb [hunk 9]
```diff
@@ -0,0 +1,118 @@
...
+    hop_run_server
+  end
+
+  label def run_server
+    case sshable.d_check(daemon_name)
+    when "InProgress"
+      hop_wait
+    when "Succeeded", "Failed", "NotStarted"
+      start_daemon
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb:72]: It seems odd to start the daemon if the status is Succeeded. Is this to handle cases where there was a previous execution that was not properly cleaned?

File: prog/storage/remote_storage_server/nexus.rb [hunk 10]
```diff
@@ -0,0 +1,118 @@
...
+    # storage is restarted, which is the intended migration semantics.
+    source = remote_storage_server.source_vm_storage_volume
+    sshable.cmd("sudo systemctl stop :unit", unit: source.vhost_backend_systemd_unit_name)
+    hop_run_server
+  end
+
+  label def run_server
+    case sshable.d_check(daemon_name)
+    when "InProgress"
+      hop_wait
```
Comment (jeremyevans) [inline prog/storage/remote_storage_server/nexus.rb:67]: This hop while InProgress is a bit unusual compared to other progs. Maybe one reason is unlike other commands, this one doesn't run and exit, but stays running. If you have other code that is waiting until this strand reaches `wait`, this seems like a reasonable approach. Otherwise, might be better to merge this `wait` label into this, and nap 30 in the in progress case, and nap 5 in the other cases.

File: rhizome/host/bin/setup-remote-storage-server [hunk 11]
```diff
@@ -0,0 +1,13 @@
+#!/usr/bin/env ruby
+# frozen_string_literal: true
+
+require "json"
+require_relative "../lib/remote_storage_server"
+
+vm_name, storage_device, disk_index, source_version, server_version, port, psk_identity = ARGV
```
Comment (jeremyevans) [inline rhizome/host/bin/setup-remote-storage-server:12]: May want to error out early if `ARGV.length != 7`

File: rhizome/host/lib/remote_storage_server.rb [hunk 12]
```diff
@@ -0,0 +1,92 @@
...
+  def listen_config_path
+    @listen_config_path ||= File.join(@sp.storage_dir, "remote-stripe-listen.conf")
+  end
+
+  def listen_config(port, psk, psk_identity)
+    [
+      toml_section("server", {"address" => "0.0.0.0:#{port}"}),
+      toml_section("server.psk", {"identity" => psk_identity, "secret.ref" => "psk"}),
+      toml_section("secrets.psk", {"source.inline" => psk, "encoding" => "base64"}),
+      toml_section("danger_zone", {"enabled" => true, "allow_inline_plaintext_secrets" => true})
```
Comment (jeremyevans) [inline rhizome/host/lib/remote_storage_server.rb]: Is `allow_inline_plaintext_secrets` as bad as it sounds? I thought the secrets were only passed over stdin and never store on disk.

File: rhizome/host/lib/remote_storage_server.rb [hunk 13]
```diff
@@ -0,0 +1,92 @@
...
+  # Run the remote-stripe-server daemon in the foreground (this process becomes
+  # the server). A forked writer streams the KEK to the volume's kek pipe, which
+  # the server reads once at startup, then this process is replaced by the
+  # server. Serving a volume whose VM is running is not supported (the vhost
+  # backend already owns the disk and the kek pipe).
+  def run(port, psk, psk_identity, kek_material)
+    fail "remote-stripe-server requires vhost block backend v0.5.0 or later" unless @server_backend.supports_remote_stripe_server?
+    write_listen_config(port, psk, psk_identity)
+
+    rm_if_exists(@sp.kek_pipe)
```
Comment (jeremyevans) [inline rhizome/host/lib/remote_storage_server.rb]: If you don't want to use a named pipe, you could use `r, w = IO.pipe`, have the forked child do `r.close` and write to the write end of the pipe (`w`), set `r.close_on_exec = false` in the parent process, and use `/dev/fd/#{r.to_i}` instead of `@sp.kek_pipe` for the exec args. One advantage of this is the transmission channel is never exposed in the file system. Up to you.

File: rhizome/host/spec/remote_storage_server_spec.rb [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+    end
+  end
+
+  describe "#run" do
+    def stub_run(srv)
+      expect(srv).to receive(:write_listen_config)
+      expect(srv).to receive(:rm_if_exists)
+      expect(File).to receive(:mkfifo)
+      expect(FileUtils).to receive(:chown)
+      expect(srv).to receive(:fork).and_return(123)
```
Comment (jeremyevans) [inline rhizome/host/spec/remote_storage_server_spec.rb]: Probably should add `with` for each of these.

File: spec/prog/storage/remote_storage_server/nexus_spec.rb [hunk 15]
```diff
@@ -0,0 +1,98 @@
...
+    it "creates a server with a psk, identity, and free port" do
+      expect(rss.source_vm_storage_volume).to eq(source_volume)
+      expect(rss.port).to be_between(5500, 5999)
+      expect(rss.psk).not_to be_nil
+      expect(rss.psk_identity).to eq(rss.ubid)
+      expect(rss.strand.label).to eq("start")
+    end
+
+    it "fails if the volume does not exist" do
+      expect { described_class.assemble(SecureRandom.uuid) }.to raise_error("No existing VmStorageVolume")
```
Comment (jeremyevans) [inline spec/prog/storage/remote_storage_server/nexus_spec.rb]: While it doesn't matter much, I think it's best to always use a correctly typed UUID for things like this:

```suggestion
      expect { described_class.assemble(VmStorageVolume.generate_uuid) }.to raise_error("No existing VmStorageVolume")
```

File: spec/prog/storage/remote_storage_server/nexus_spec.rb [hunk 16]
```diff
@@ -0,0 +1,98 @@
...
+    end
+  end
+
+  describe "#destroy" do
+    it "stops the daemon and destroys the model" do
+      expect(sshable).to receive(:d_stop).with(nx.daemon_name)
+      expect(sshable).to receive(:d_clean).with(nx.daemon_name)
+      id = rss.id
+      expect { nx.destroy }.to exit({"msg" => "remote storage server destroyed"})
+      expect(RemoteStorageServer[id]).to be_nil
```
Comment (jeremyevans) [inline spec/prog/storage/remote_storage_server/nexus_spec.rb]:
```suggestion
      expect(rss).not_to exist
```

File: prog/storage/remote_storage_server/nexus.rb (hunk 5 above)
Comment (jeremyevans) [reply prog/storage/remote_storage_server/nexus.rb]: I checked and the Linux ephemeral port range is 32768 to 60999, so I guess we can assume the port is available.

File: prog/storage/remote_storage_server/nexus.rb (hunk 6 above)
Comment (jeremyevans) [reply prog/storage/remote_storage_server/nexus.rb]: I decided to avoid this issue and add `remote_storage_server.vm_host_id`, as that was necessary to get a unique constraint on `[vm_host_id, port]`, which we definitely want.

File: prog/storage/remote_storage_server/nexus.rb (hunk 7 above)
Comment (jeremyevans) [reply prog/storage/remote_storage_server/nexus.rb]: `Prog::Base#before_run` has the same behavior if the `destroy` semaphore is set, so I think we can remove this method and the related spec.

File: prog/storage/remote_storage_server/nexus.rb (hunk 8 above)
Comment (jeremyevans) [reply prog/storage/remote_storage_server/nexus.rb:60]: The expected immediate use for the remote storage server is to move VM storage from one host to another, and the prerequisite for that will be the VM is in prepare to move state (which includes being at the admin_stop label), even before we call `Prog::Storage::RemoteStorageServer::Nexus.assemble`. So I don't think we need a separate check here.

File: prog/storage/remote_storage_server/nexus.rb (hunk 9 above)
Comment (jeremyevans) [reply prog/storage/remote_storage_server/nexus.rb:72]: Going to add a `Clog.emit` for the Succeeded state, since that case is unexpected.

File: prog/storage/remote_storage_server/nexus.rb (hunk 10 above)
Comment (jeremyevans) [reply prog/storage/remote_storage_server/nexus.rb:67]: I determined the reason for this is the `register_deadline` in start. If the server never starts correctly, that will result in a page. It wouldn't handle cases where the server process stops after starting and cannot be restarted, so I'll add a deadline registration in wait for jumping back to wait.

File: rhizome/host/lib/remote_storage_server.rb (hunk 12 above)
Comment (jeremyevans) [reply rhizome/host/lib/remote_storage_server.rb]: Rebased and pushed an approach that passes the listen config over stdin, so it's never unencrypted on disk. Added a comment explaining that.

File: rhizome/host/lib/remote_storage_server.rb (hunk 13 above)
Comment (jeremyevans) [reply rhizome/host/lib/remote_storage_server.rb]: While it appears the kek_pipe code is only used in the legacy case, it's actually used in the non-legacy case as well, as part of the config file. After discussions with Hadi, this code could be simplified using `run_with_kek_pipe`, so I used that approach.

File: rhizome/host/spec/remote_storage_server_spec.rb (hunk 14 above)
Comment (jeremyevans) [reply rhizome/host/spec/remote_storage_server_spec.rb]: Avoided this issue by having the expectation on the arguments passed to `run_with_kek_pipe`.

**Comment by jeremyevans** [comment]
/run-e2e metal move_vm

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/30674819717

**Comment by jeremyevans** [comment]
E2E test failure due to lack of boot image. I'm guessing this is because I attempted to run the test in isolation, and didn't add the necessary code to download the boot image first.

**Comment by jeremyevans** [comment]
/run-e2e metal

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/30854799587

**Comment by jeremyevans** [comment]
Last E2E run failed before even getting to MoveVm:

```
22:30:38 e2e.1       | 59.86m | d30c5273-7494-8b3a-8c02-5aaf6794e2c6 | Test::VmGroup.verify_vms_after_reboot | waiting for exit | Vm.wait, Vm.wait, Vm.wait, Vm.wait
```

**Comment by jeremyevans** [comment]
/run-e2e metal

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/30871232807

**Comment by jeremyevans** [comment]
E2E passed, but it does not appear to have run the MoveVm tests. I thought that without a test case given, it would run all test cases, but apparently it does not.

**Comment by jeremyevans** [comment]
/run-e2e metal vm,move_vm

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/30884964566

**Comment by jeremyevans** [comment]
E2E failure was the same boot image failure during allocation that we had when running `move_vm` in isolation: `[["base",1],["space",1],["boot_image",0]]`.

**Comment by jeremyevans** [comment]
Determined I needed to update the allocator to account for the remote storage server and not require a boot image (similarly to how machine images are currently handled).

**Comment by jeremyevans** [comment]
/run-e2e metal move_vm

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/30939289506

**Comment by jeremyevans** [comment]
/run-e2e metal move_vm

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/30949615056

**Comment by jeremyevans** [comment]
Last E2E test failure was because the new Vm didn't have an sshable created.  Added a new commit to create an sshable for new Vm if current Vm has an sshable.

**Comment by jeremyevans** [comment]
/run-e2e metal move_vm

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/30958737983

**Comment by jeremyevans** [comment]
E2E passed, marking as ready for review. @pykello It doesn't let me request your review, since you were the initial submitter, but I'd like you to review.

File: config/e2e_test_cases.yml [hunk 17]
```diff
@@ -11,3 +11,4 @@
 - { name: kubernetes_upgrade,        images: ["ubuntu-noble", "kubernetes-v1_35", "kubernetes-v1_36"] }
 - { name: kubernetes_firewall,       images: ["ubuntu-noble", "kubernetes-v1_35"] }
 - { name: machine_image_vm_capture,  images: ["ubuntu-noble"] }
+- { name: move_vm,                   images: ["ubuntu-noble"] }
```
Comment (pykello) [inline config/e2e_test_cases.yml:14]: Do we also need to update `test_cases` in `.github/workflows/e2e.yml`?

File: prog/storage/remote_storage_server/nexus.rb [hunk 18]
```diff
@@ -0,0 +1,117 @@
...
+  # The server always runs the remote-stripe-server binary, which can
+  # serve volumes created by older backends.
+  SERVER_VERSION = "v0.5.0"
+
+  # Given the volume to serve, figure out its host, pick a free port on that
+  # host, create a PSK, and start the server.
+  def self.assemble(vm_storage_volume_id)
+    source_volume = VmStorageVolume[vm_storage_volume_id]
+    fail "No existing VmStorageVolume" unless source_volume
+    fail "Source volume must be encrypted" unless source_volume.key_encryption_key_1
```
Comment (pykello) [inline prog/storage/remote_storage_server/nexus.rb:22]: I think we should also verify that:
1. the server has ubiblk version `SERVER_VERSION`
2. VM is stopped

File: prog/storage/remote_storage_server/nexus.rb [hunk 19]
```diff
@@ -0,0 +1,117 @@
...
+    hop_run_server
+  end
+
+  label def run_server
+    case state = sshable.d_check(daemon_name)
+    when "InProgress"
+      hop_wait
+    when "Failed", "NotStarted"
+      start_daemon
+    else # when "Succeeded"
```
Comment (pykello) [inline prog/storage/remote_storage_server/nexus.rb]:
```suggestion
    else
```

File: prog/test/move_vm.rb [hunk 20]
```diff
@@ -0,0 +1,104 @@
...
+  semaphore :destroy
+
+  MARKER_COUNT = 5
+
+  def self.assemble(vm_host, project_id: nil)
+    DB.transaction do
+      project_id ||= Project.create(name: "move-vm-test").id
+
+      if vm_host.vhost_block_backends_dataset.where { version_code >= 500 }.empty?
+        Prog::Storage::SetupVhostBlockBackend.assemble(VmHost.first.id, "v0.5.0", allocation_weight: 100)
```
Comment (pykello) [inline prog/test/move_vm.rb:16]: We will eventually want to move all hosts to `v0.5.0`, so we could instead change `VHOST_BLOCK_BACKEND_VERSION` to `v0.5.0` in `.github/workflows/e2e.yml` to start testing the new version for broader uses.

File: prog/vm/metal/move_vm.rb [hunk 21]
```diff
@@ -0,0 +1,73 @@
...
+
+class Prog::Vm::Metal::MoveVm < Prog::Base
+  subject_is :vm
+
+  frame_reader :remote_storage_server_id
+
+  def self.assemble(vm, vm_host, parent_id: nil)
+    fail "Vm is not ready to move" unless vm.strand.label == "stopped_by_admin" && vm.prepare_to_move_set?
+    fail "VmHost is not in the same location as the Vm" unless vm_host.location_id == vm.location_id
+    fail "Vm must have a single storage volume" unless vm.vm_storage_volumes.count == 1
```
Comment (pykello) [inline prog/vm/metal/move_vm.rb:12]: if VM is burstable, then `vm_host.accepts_slices` must also be true.

This will be true for all new hosts and for most of old hosts, but we still have few old hosts with no slice support.

File: prog/vm/metal/move_vm.rb [hunk 22]
```diff
@@ -0,0 +1,73 @@
...
+
```
Comment (pykello) [inline prog/vm/metal/move_vm.rb:14]: Although we probably won't move `arm64` VMs, but it might be worth adding a check to verify that the target host arch is equal to vm arch as a precaution.

File: spec/prog/vm/metal/move_vm_spec.rb [hunk 23]
```diff
@@ -0,0 +1,212 @@
...
+      new_volume = new_vm.vm_storage_volumes.first
+      expect(new_volume.size_gib).to eq(old_volume.size_gib)
+      expect(new_volume.remote_storage_server_id).to eq(rss.id)
+      expect(new_vm.strand.stack.first["force_host_id"]).to eq(vm_host.id)
+
+      expect(st.prog).to eq("Vm::Metal::MoveVm")
+      expect(st.label).to eq("start")
+      expect(st.stack.first["subject_id"]).to eq(new_vm.id)
+      expect(st.stack.first["remote_storage_server_id"]).to eq(rss.id)
+    end
```
Comment (pykello) [inline spec/prog/vm/metal/move_vm_spec.rb:127]: it might be worth to also test the path where parent_id is not nil, possibly by modifying one of the existing happy-path specs.

File: prog/test/move_vm.rb (hunk 20 above)
Comment (jeremyevans) [reply prog/test/move_vm.rb:16]: Agree on the `VHOST_BLOCK_BACKEND_VERSION` change. However, we still want this check, because this test is designed to be runnable locally (that's why it accepts a `project_id` keyword argument), and the host may not have the 0.5.0 backend installed.

**Comment by jeremyevans** [comment]
@pykello I've pushed additional commits to address all of your commits. amd64 CI was consistently segfaulting due to Aws::EC2::Client loading with this branch, so I also pushed a commit to change the location of the load, which appears to have worked around the issue. Hopefully this issue will be fixed in Ruby 4.0.7 (I think it is fixed in Ruby 4.0.6, but with a severe performance regression).

**Comment by jeremyevans** [comment]
/run-e2e metal

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/31051900012

**Comment by jeremyevans** [comment]
E2E passed with a full metal run, including the move_vm test (now included in the default).

**Review by jeremyevans — APPROVED** [review]
Approving this as @pykello was the originally submitter and cannot officially approve it. I'll wait for @pykello to comment with approval before merging this.

File: prog/storage/remote_storage_server/nexus.rb [hunk 24]
```diff
@@ -0,0 +1,121 @@
...
+  # Given the volume to serve, figure out its host, pick a free port on that
+  # host, create a PSK, and start the server.
+  def self.assemble(vm_storage_volume_id)
+    source_volume = VmStorageVolume[vm_storage_volume_id]
+    fail "No existing VmStorageVolume" unless source_volume
+    fail "Source volume must be encrypted" unless source_volume.key_encryption_key_1
+    vm = source_volume.vm
+    vm_host = vm.vm_host
+    fail "VM isn't in stopped_by_admin state" unless vm.strand.label == "stopped_by_admin"
+    fail "Host doesn't have ubiblk #{SERVER_VERSION}+" if vm_host.vhost_block_backends_dataset.where { version_code >= 500 }.empty?
```
Comment (pykello) [inline prog/storage/remote_storage_server/nexus.rb]: Here we check for `version_code >= 500` but in `start_daemon` we pass exact `SERVER_VERSION`. I think we should either be flexible there or do an exact match here.

**Review by pykello — COMMENTED** [review]
Looks good & I think we can merge this and do incremental improvements/fixes on top of it.

In my tests, I ran into the following issues, which I think can be fixed after merging this.
- Legacy move didn't reach `wait`. The ubicloud setup seemed correct, so this is probably ubiblk issue & I can debug it after we merge this.
- Removing VM while the move is running causes exceptions in `nap 30 vm.strand.label == "wait"`. We should probably prevent VM destroy while the move is running. I think this could also be fixed after merging this.

So, I think this can be merged after addressing the comment about `SERVER_VERSION`. (Since I have created this PR, I can't "Approve" it using github).

File: prog/storage/remote_storage_server/nexus.rb (hunk 24 above)
Comment (jeremyevans) [reply prog/storage/remote_storage_server/nexus.rb]: I think we should be flexible. Pushed another commit for that.

**Comment by jeremyevans** [comment]
/run-e2e metal

E2E tests triggered: https://github.com/ubicloud/ubicloud/actions/runs/31444452713
